### PR TITLE
allow `autorun()`

### PR DESCRIPTION
**Pull request by KATT**
was curious about normi and did a test to see how it works if one were to update 

a) if updating a child node would make the parent node would "react"
b) if updating an unrelated node would trigger observe

---

seems `makeObservable(this)` is needed in `Normi`  constructor for it to work properly

### CODE REVIEW

File: src/store.ts [hunk 1]
```diff
@@ -11,6 +11,7 @@ export class Normi {
     this.params = {
       id: params.id || ['id'],
     };
+    makeObservable(this);
```
Comment (colinhacks) [inline src/store.ts]: Instead, I recommend just using `makeAutoObservable` and getting rid of all the decorators in the class definition. When I wrote `normi` initially, makeAutoObservable didn't exist.

**Comment by colinhacks** [comment]
I'm having trouble getting the tests to work after having previously merged #3. Can you try to merge this branch with the new HEAD locally and update this PR?

Screen Shot 2021-01-31 at 9 38 19 PM

**Comment by KATT** [comment]
@colinhacks this has now been updated

One access-pattern that got me a bit puzzled is how one is best to deal with args on child nodes, for example accessing a user that has a paginated list of `posts` underneath.

Guess one would have to format the response something like this:

```ts

const response = {
  user: {
    id: '...',
    name: 'colin',
  },
  userPosts: [
    {
      id: '...',
      title: 'first post',
    },
  ],
};
```

but it gets a bit unintuitive.. thinking a bit about a good access pattern for this in conjunction with tRPC. GraphQL has the arguments for each node embedded in the cache key, but that wouldn't really work here. Would be great with ideas so we know how to inform people to do common access patterns like this. This is one of the things that would stop us from being able to simply bundle normi with `@trpc/react` as it wouldn't _always_ work and leave users with nasty bugs.

**Comment by colinhacks** [comment]
Not sure I understand the problem.

Overwriting response.userPosts each time is fine as far as I'm concerned, as long as the old `userPosts` are each still stored in the cache. 

The old version of tRPC supported putting the args in the cache key by splitting up each call into an initialization step and an execution step:

```ts

const mutation = SDK.org.create(data);
mutation.payload.args; // unknown[]
mutation.payload.path; // string[]
mutation.payload.context; // unknown
const result = await mutation.run();
```

After initializing the query/mutation you can access all the metadata about it (path,context,args) and use it to generate a cache_key upfront. 

Normi probably shouldn't be bundled in by default anyway, since it involves a dependency on mobx. Most people don't want/need that. That's why I tried to keep the original tRPC as lean as possible, and users could implement their own fetching logic when they initialize the SDK. 

```ts
const SDK = trpc.sdk<RootRouter>({
  url: 'http://localhost:3000/api/rpc',
  getContext: () => {
    return {};
  },
  handler: async (url, payload) => {
    try {
      const result = await Axios.post(url, {
        ...payload,
        context: { token },
      });
      return normi.merge(result.data).value;
    } catch (err) {
      console.log(err);
      throw err;
    }
  },
});
```

Providing a basic react hook out of the box is still a good idea but you don't want to make it too opinionated.

**Comment by KATT** [comment]
> The old version of tRPC supported putting the args in the cache key by splitting up each call into an initialization step and an execution step:

The http client is it's own package & it's still possible to make it intercept-able or make another client that is.


> Normi probably shouldn't be bundled in by default anyway, since it involves a dependency on mobx. Most people don't want/need that. That's why I tried to keep the original tRPC as lean as possible, and users could implement their own fetching logic when they initialize the SDK.

💯 

I started thinking about these things just because I want to use it. It'll works great as long as one puts the properties that are affected by a variable outside of the actual object. 

All I was saying really is that it's tempting to do something like 

```ts
const response = {
  user: {
    id: '...',
    name: 'colin',
    posts: [ /* */ ],
  },
};
```

But that won't work well in case `posts` is paginated as different responses would overwrite the cache.


---

Anyway, could you review this PR? :)

**Pull request merged**